src/dbnode/storage/index/compaction/plan.go

```diff
@@ -0,0 +1,240 @@
+
+	// group segments into buckets (2a)
+	segmentsByBucket := make(map[SizeBucket][]Segment, len(buckets))
+	for _, seg := range compactableSegments {
+		var (
+			bucket      SizeBucket
+			bucketFound bool
+		)
+		for _, b := range buckets {
+			if b.MinSizeInclusive <= seg.Size && seg.Size < b.MaxSizeExclusive {
+				bucket = b
+				bucketFound = true
+				break
+			}
+		}
+		if !bucketFound {
+			plan.UnusedSegments = append(plan.UnusedSegments, seg)
+		}
+		segmentsByBucket[bucket] = append(segmentsByBucket[bucket], seg)
+	}
+
+	// for each bucket, sub-group segments into tier'd sizes (2b)
+	for bucket, bucketSegments := range segmentsByBucket {
+		var (
+			task            Task
+			accumulatedSize int64
+		)
+		sort.Slice(bucketSegments, func(i, j int) bool {
+			return bucketSegments[i].Size < bucketSegments[j].Size
+		})
+		for _, seg := range bucketSegments {
+			accumulatedSize += seg.Size
+			task.Segments = append(task.Segments, seg)
+			if accumulatedSize >= bucket.MaxSizeExclusive {
+				plan.Tasks = append(plan.Tasks, task)
+				task = Task{}
+				accumulatedSize = 0
+			}
+		}
+		// fall thru cases: no accumulation, so we're good
+		if len(task.Segments) == 0 || accumulatedSize == 0 {
+			continue
+		}
+
+		// in case we never went over accumulated size, but have 2 or more segments, we should still compact them
+		if len(task.Segments) > 1 {
+			plan.Tasks = append(plan.Tasks, task)
+			continue
+		}
+
+		// even if we only have a single segment, if its a mutable segment, we should compact it to convert into a FST
+		if task.Segments[0].Type == segments.MutableType {
+			plan.Tasks = append(plan.Tasks, task)
+			continue
+		}
+
+		// at this point, we have a single FST segment but don't need to compact it; so mark it as such
+		plan.UnusedSegments = append(plan.UnusedSegments, task.Segments[0])
+	}
+
+	// now that we have the plan, we priortise the tasks as requested in the opts. (2c)
+	sort.Stable(plan)
+	return plan, nil
+}
+
+func (p *Plan) Len() int      { return len(p.Tasks) }
+func (p *Plan) Swap(i, j int) { p.Tasks[i], p.Tasks[j] = p.Tasks[j], p.Tasks[i] }
+func (p *Plan) Less(i, j int) bool {
+	switch p.OrderBy {
+	case TasksOrderedByOldestMutableAndSize:
+		fallthrough
+	default:
+		taskSummaryi, taskSummaryj := p.Tasks[i].Summary(), p.Tasks[j].Summary()
+		if taskSummaryi.CumulativeMutableAge != taskSummaryj.CumulativeMutableAge {
+			// i.e. put those tasks which have cumulative age greater first
+			return taskSummaryi.CumulativeMutableAge > taskSummaryj.CumulativeMutableAge
+		}
+		if taskSummaryi.NumMutable != taskSummaryj.NumMutable {
+			// i.e. put those tasks with more mutable segments first
+			return taskSummaryi.NumMutable > taskSummaryj.NumMutable
+		}
+		// i.e. smaller tasks over bigger ones
+		return taskSummaryi.CumulativeSize < taskSummaryj.CumulativeSize
+	}
+}
+
+func (o PlannerOptions) Validate() error {
+	if o.MutableCompactionAge <= 0 {
+		return errMutableCompactionAgeNegative
```

Comment:
Hm, we should probably allow compaction age of zero though no?  Maybe just reduce the check to `< 0`?

Reply:
Sure thing, don't think it'd ever happen but definitely more correct to say `< 0`

Reply:
done